src/main/java/org/jenkinsci/plugins/durabletask/executors/OnceRetentionStrategy.java

```diff
@@ -71,7 +71,19 @@ public long check(final AbstractCloudComputer c) {
                 done(c);
             }
         }
-
+        // if the agent is not done and it is offline we should trigger a relaunch
+        // do not do this whilst we hold the lock, as connect is documented only as "This method may return immediately"
+        // so lets not presume any asynchronous nature and instead that this could be a long blocking op.
+        boolean shouldRestart = false;
+        synchronized (this) {
```

Comment:
whilst this locking is not perfect (we may try to relaunch and agent just
as it was done), we have set the computer to not accept any tasks so the
agent should become idle, and be terminated (either in the current
termination block) or the next time the check runs.